From 99912f375bcf473ab25d2d3e14982d68d64d8776 Mon Sep 17 00:00:00 2001
From: Jiri Pirko <jpirko@redhat.com>
Date: Mon, 15 Sep 2014 09:32:05 -0400
Subject: [net] vti: Use the tunnel mark for lookup in the error handlers

Message-id: <1410773525-8735-3-git-send-email-jpirko@redhat.com>
Patchwork-id: 94793
O-Subject: [patch rhel7 net 21/21] vti: Use the tunnel mark for lookup in the error handlers.
Bugzilla: 1091561
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ1091561
https://brewweb.devel.redhat.com/taskinfo?taskID=7954687

upstream commit 6d004d6cc73920299adf4cfe25010b348fc94395
Author: Steffen Klassert <steffen.klassert@secunet.com>
Date:   Mon May 12 09:09:26 2014 +0200

    vti: Use the tunnel mark for lookup in the error handlers.

    We need to use the mark we get from the tunnels o_key to
    lookup the right vti state in the error handlers. This patch
    ensures that.

Signed-off-by: Jiri Pirko <jpirko@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/ip_vti.c b/net/ipv4/ip_vti.c
index b6d7f97..d855c69 100644
--- a/net/ipv4/ip_vti.c
+++ b/net/ipv4/ip_vti.c
@@ -239,6 +239,7 @@ static netdev_tx_t vti_tunnel_xmit(struct sk_buff *skb, struct net_device *dev)
 static int vti4_err(struct sk_buff *skb, u32 info)
 {
  __be32 spi;
+ __u32 mark;
  struct xfrm_state *x;
  struct ip_tunnel *tunnel;
  struct ip_esp_hdr *esph;
@@ -254,6 +255,8 @@ static int vti4_err(struct sk_buff *skb, u32 info)
  if (!tunnel)
   return -1;
 
+ mark = be32_to_cpu(tunnel->parms.o_key);
+
  switch (protocol) {
  case IPPROTO_ESP:
   esph = (struct ip_esp_hdr *)(skb->data+(iph->ihl<<2));
@@ -281,7 +284,7 @@ static int vti4_err(struct sk_buff *skb, u32 info)
   return 0;
  }
 
- x = xfrm_state_lookup(net, skb->mark, (const xfrm_address_t *)&iph->daddr,
+ x = xfrm_state_lookup(net, mark, (const xfrm_address_t *)&iph->daddr,
          spi, protocol, AF_INET);
  if (!x)
   return 0;
-- 
1.7.1